google/cloud/storage/object_rewriter.h

```diff
@@ -90,15 +90,16 @@ class ObjectRewriter {
    *
    * @return the object metadata once the copy completes.
    */
-  template <typename Functor,
-            typename std::enable_if<google::cloud::internal::is_invocable<
-                                        Functor, RewriteProgress>::value,
-                                    int>::type = 0>
-  ObjectMetadata ResultWithProgressCallback(Functor cb) {
-    while (not progress_.done) {
+  template <
+      typename Functor,
+      typename std::enable_if<google::cloud::internal::is_invocable<
+                                  Functor, StatusOr<RewriteProgress>>::value,
+                              int>::type = 0>
+  StatusOr<ObjectMetadata> ResultWithProgressCallback(Functor cb) {
+    do {
```

Comment:
This looks like a change in behavior if called when progress_->done() is already true.  Intentional?

Reply:
Yes, but I might have missed something. Let me know if this makes sense:

- It used to be that `progress_` was just a struct, with `progress_.done` initialized to false so the loop was always entered and continued to run until `done` was true.
- I thought that `progress_` needed to become a `StatusOr<>` because on failure we want to communicate that to the progress callback and whatever called this function.
- So the only way to *use* `progress_` is to say something like `progress_.ok() and not progress_->done`.
- If we just write `while (progress_.ok() and progress_->done)` the loop never enters, that is bad, a "change" in behavior as "it no longer does its work".
- So we need to check *after* calling Iterate.
- So I convinced myself a do/while was the "Right Thing":tm: : The loop enters at least once (like before), only then we check the value.
- Both before and now behave weird if you call this loop twice.

So after writing that, maybe what we should do is
- keep `progress_` as a struct,
- store the error status in a separate field,
- set `progress_.done` to `true` on the first error,
- and return either the `status_` field or the `progress_` field depending on whether there was an error.

I think that looks cleaner (sending code in a few minutes).

A separate question is what to do when an error is returned: should we invoke the callback?  We used to not, because the exception would propagate out, but I thought we wanted the application developer to see those errors and they can decide what do with the error.